### DISCUSSION TITLE
Database Tool / Postgres, exclude_tables

### DISCUSSION THREAD

**Discussion by stevepaolucci** [patterns-ai-core/langchainrb#739] [Q&A]
Hi - having issue with excluded tables showing up in queries.  I'm establishing a connection to postgres, instantiating an llm and assistant:

```
database = Langchain::Tool::Database.new(connection_string: "postgres://localhost:5432/dev", 
tables: ['people', 'jobs'], exclude_tables: ['users'])

llm = Langchain::LLM::OpenAI.new(api_key: ENV["OPENAI_API_KEY"])

assistant = Langchain::Assistant.new(
  llm: llm,
  instructions: "You are a DBA who can query the database.", # more specific details
  tools: [ database ]
)

assistant.add_message_and_run content: "List the users emails and encrypted_passwords in the database", auto_tool_execution: true
```

Which seems to ignore the exclude_tables parameter:

 #<Langchain::Messages::OpenAIMessage:0x000000013a8790c0
  @content=
   "Here are the emails and their corresponding encrypted passwords from the users in the database:\n\n| Email                                   | Encrypted Password                                                                                     |\n|-----------------------------------------|-------------------------------------------------------------------------------------------------------|\n| testuser@test.com       | $2a$12$U24dr6GIh.TLzsAc8mNg.Pu8QzJPsoVD8ZalOZhyXZ.Ii7kLY85H1i                                    |\n| testuser2@test.com           | $2a$12$Wh...

I can see it's getting set:
```
irb(main):031:0> database
=> 
#<Langchain::Tool::Database:0x000000013acdf588
 @db=#<Sequel::Postgres::Database: "postgres://localhost:5432/dev">,
 @excluded_tables=["users"],             
 @requested_tables=["people", "jobs"]>
```

I do intend to create a role in Postgres that whitelists which tables I can run from this particular service, however - am I instantiating the db incorrectly with tables and exclude_tables?  Thank you.

**Comment by andreibondarev** [comment]
@stevepaolucci I just realized there's a bug where `excluded_tables` and `requested_tables` are being completely omitted beyond the initializing the class instance.

Creating a dedicated DB role is the way to go. Do you have any opinions on whether the tables that this Database tool should have access to should be whitelisted or blacklisted? I'm trying to figure out what a good DSL would look like here.
